### ISSUE TITLE
Rolling release / continuous deployment

### ISSUE THREAD

**Issue by 9999years** [topgrade-rs/topgrade#222]
The current approach of basing changes off of `dev`, which is then occasionally merged into `master`, adds an additional layer of complexity that I hope we can eliminate. What if, instead, we were to make PRs to `master` and have CI build and publish a release on basically every commit? We'll publish a lot of versions but that's fine, we have a lot of patches and people don't have to upgrade if they don't need to.

(I have such a system, which builds and publishes a release on every commit in GitHub Actions, and automatically bumps version numbers based on PR labels. I can PR it and help set it up if you'd like.)

**Comment by DottoDev** [comment]
This would be great if we say that bug fixes will be merged into master and new features will be merged into dev. This would result on still being  semver compliant.

**Comment by 9999years** [comment]
Hm? Bug fixes can be merged into master as patch-level version bumps and new features can be merged into master as minor-level version bumps.

**Comment by DottoDev** [comment]
I would like it more if we have a dedicated branch for new-features.
Every day a CI-pipeline runs which checks if there are new features available, if yes, merge with master and create a new minor release. If there isn't a change in new-features it checks if there is just a change on the master branch, if yes it creates a bug fix release.
This is ust my idea but it should decide on it's own if it's a a bug-fix or minor release.

**Comment by Akeshihiro** [comment]
With two branches you have to merge accepted bugfix PRs into master also into the dev branch, otherwise both branches diverge and the more they diverge, the more issues could come up because of the differences. So there would be some effort needed to sync both branches.

With workflows like git-flow this is no issue, but not everybode knows such workflows and just drop some PRs on some branches, mostly on the master branch. Personlly, I like the git-flow because of the fact that you have two or more long living branches (f.e. master/main and develop) and generally everything goes into the develop branch, not the master branch. The master branch signals the most recent stable release. So new development goes into the develop branch until someone decides it's time for a new release. So bugfixes generally also are merged into develop. An exception to this are hotfixes. Hotfix branches are based on the master branch instead of the develop branch and are merged into the master branch as well as into the develop branch, so that the develop branch gets the hotfix as well and stays in sync that way. This does not need to be made manually, there are some git-flow skripts/tools out there, so that you don't have to know all this "ceremony", you can just use the git-flow scripts/tools (which can be seen as git extensions).

That being said, I also think that two branches generally are a good thing. A hotfix (propably a security issue fix), which needs to be deployed ASAP, is a completely another story than new features or minor bugfixes. But if you deploy every new commit and there each new feature and bugfix and what not, then you may deploy broken software and I really don't like that idea. Even on Linux distros like Arch Linux there's some kind of first testing phase (the unstable branch), so even on Arch Linux nobody gets the newest and untested stuff. I like the idea of merging hotfixes (and maybe bugfixes as well) into master and automating the version incrementation and deploying that stuff. But for new features that's really another story. I don't want to rebuild and update topgrade everyday or worse multiple times a day (I installed topgrade via cargo) just because of such automation. My machine would have to recompile the whole thing multiple times a day, wtf. Such a waste of electricity and time.  Well, the latter case would be my personal issue that I would have to solve then somehow (ignoring topgrade updates and upgrade topgrade only once a week or something like that), but on the other had I think that that's not only my personal problem, but of all of us because of the whole process going on in the background (building, testing, creating and deploying a release and also deploying that release to crates.io or who knows where else) and consuming all that computation power and electricity just because one PR got merged ... Well ...

But this are just my 2 cents, so  do whatever you want. But I don't like the idea that such a little (very useful tool, for sure) as topgrade should have such a big footprint. So yeah, the stage is yours :D

**Comment by 9999years** [comment]
> I like the git-flow because of the fact that you have two or more long living branches (f.e. master/main and develop) and generally everything goes into the develop branch, not the master branch. The master branch signals the most recent stable release.

I think git tags / GitHub releases are a better way to signal 'most recent stable' than the most recent `master` commit.

> But if you deploy each new commit (accepted PR) and each new feature and bugfix and what not, then you may deploy broken software and I really don't like that idea

This is always a possibility, and in my opinion is generally more likely when you're releasing multiple changes at a time in a bundle. Smaller changes are less likely to break things, and easier to revert in case of problems. In a worst-case scenario we can even yank (revoke) releases. 

> I don't want to rebuild and update topgrade everyday or even worse multiple times a day (I installed topgrade via cargo)

Don't do that, then. Use a package manager. The git repo is for development. We ship compiled binaries for a reason. I do not want to optimize our development workflow for comfort with `cargo install`, I want it to be painless to build and publish releases.

**Comment by Akeshihiro** [comment]
> I think git tags / GitHub releases are a better way to signal 'most recent stable' than the most recent `master` commit.

> This is always a possibility, and in my opinion is generally more likely when you're releasing multiple changes at a time in a bundle. Smaller changes are less likely to break things, and easier to revert in case of problems. In a worst-case scenario we can even yank (revoke) releases.

Well, that might be true if you prefer trunk based development and that's totally fine and I am fine with that as well. But that's not my point here. My point is that if you automate the deployment based on an accepted PR, I think then not only the PR will be merged but also a new release would be built (creating a tag and a release) and deployed (to crates.io and who knows what else). And that's my issue here. If there will be created and deployed a new release for each accepted little PR, then you push a lot of releases into the world. It doesn't matter how the people install topgrade, like by using `cargo install` (which is as far as I know pretty much normal to install rust apps) or by using any other package manager. For example AUR on Arch Linux also builds topgrade and as far as I know this is true for pretty much all other package managers as well, except for brew on MacOS maybe. And even if the package managers would install a prebuilt binary, that would not change anything because of the frequency of the deployments.

Maybe I missunderstood something and if so, I am sorry. But as I understood this proposal, there would be a release created for each accepted PR. Am I wrong with this? If so, then forget my interruption on that. But if I understood it right, I must say that I do not like the idea to automatically create a release for each PR. I propose a more milestone like release procedure, like collecting some PRs and then creating a release if there's enough stuff that went in.
